edb/edgeql/compiler/stmtctx.py

```diff
@@ -818,6 +818,15 @@ def check_params(params: Dict[str, irast.Param]) -> None:
                 f'{"s" if len(missing_args) > 1 else ""}')
 
 
+def throw_on_shaped_param(
+    param: qlast.Parameter,
+    ctx: context.ContextLevel
+) -> None:
+    raise errors.QueryError(
+        f'a shape cannot be applied to the parameter',
+        context=param.context)
```

Comment:
Lets add a `hint=` explaining that probably parentheses should be added around the cast+param